Return WS error code & reason on closure due to balance threshold exceeded

movitto opened the pull request:
Clients can now inspect error code & msg to ascertain reason for connection being closed.

Fixes issues #2883 and #2887

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

ripplelabs-jenkins commented:
## Jenkins Build Summary

Built from [this commit](https://github.com/ripple/rippled/commit/007cd90b158d47f7b48c5f277a7b9f473c7f3fc1)

Built at __20190522 - 04:34:54__

### Test Results

Build Type | Log | Result | Status
---------- | --- | ------ | ------
rpm | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/rpm.txt) | 1177 cases, 0 failed, t: n/a | PASS :white_check_mark: 
msvc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/msvc.Debug.txt) | 1177 cases, 0 failed, t: 619s | PASS :white_check_mark: 
msvc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/msvc.Debug_NINJA_BUILD_true.txt) | 1177 cases, 0 failed, t: 639s | PASS :white_check_mark: 
msvc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/msvc.Debug_Dunity_OFF.txt) | 1176 cases, 0 failed, t: 1105s | PASS :white_check_mark: 
msvc.Release | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/msvc.Release.txt) | 1177 cases, 0 failed, t: 422s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON, <br/>MANUAL_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Release_Dassert_ON_MANUAL_TESTS_true.txt) | 915 cases, 0 failed, t: 5m15s | PASS :white_check_mark: 
docs, <br/>TARGET=docs | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/docs_TARGET_docs.txt) | 1 cases, 0 failed, t: 0m1s | PASS :white_check_mark: 
gcc.Debug <br/>-Dcoverage=ON, <br/>TARGET=coverage_report, <br/>SKIP_TESTS=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Debug_Dcoverage_ON_TARGET_coverage_report_SKIP_TESTS_true.txt) | 1177 cases, 0 failed, t: 16m44s | PASS :white_check_mark: 
clang.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/clang.Debug.txt) | 1177 cases, 0 failed, t: 2m59s | PASS :white_check_mark: 
gcc.Debug | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Debug.txt) | 1177 cases, 0 failed, t: 3m7s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/clang.Debug_Dunity_OFF.txt) | 1176 cases, 0 failed, t: 10m26s | PASS :white_check_mark: 
clang.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/clang.Release_Dassert_ON.txt) | 1177 cases, 0 failed, t: 4m59s | PASS :white_check_mark: 
gcc.Debug <br/>-Dunity=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Debug_Dunity_OFF.txt) | 1176 cases, 0 failed, t: 11m33s | PASS :white_check_mark: 
gcc.Release <br/>-Dassert=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Release_Dassert_ON.txt) | 1177 cases, 0 failed, t: 5m10s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Debug_Dstatic_OFF.txt) | 1177 cases, 0 failed, t: 3m4s | PASS :white_check_mark: 
gcc.Debug <br/>-Dstatic=OFF -DBUILD_SHARED_LIBS=ON | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Debug_Dstatic_OFF_DBUILD_SHARED_LIBS_ON.txt) | 1177 cases, 0 failed, t: 3m1s | PASS :white_check_mark: 
gcc.Debug, <br/>NINJA_BUILD=true | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/gcc.Debug_NINJA_BUILD_true.txt) | 1177 cases, 0 failed, t: 3m3s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=address, <br/>PARALLEL_TESTS=false, <br/>DEBUGGER=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/clang.Debug_Dunity_OFF_Dsan_address_PARALLEL_TESTS_false_DEBUGGER_false.txt) | 1176 cases, 0 failed, t: 10m49s | PASS :white_check_mark: 
clang.Debug <br/>-Dunity=OFF -Dsan=undefined, <br/>PARALLEL_TESTS=false | [logfile](http://ci.ops.ripple.com/job/rippled-dev/job/PR-2948/1//artifact/clang.Debug_Dunity_OFF_Dsan_undefined_PARALLEL_TESTS_false.txt) | 1176 cases, 0 failed, t: 12m49s | PASS :white_check_mark:

HowardHinnant commented on src/ripple/server/impl/BaseWSPeer.h:
```diff
@@ -234,14 +237,36 @@ close()
     if(! strand_.running_in_this_thread())
         return strand_.post(std::bind(
-            &BaseWSPeer::close, impl().shared_from_this()));
+            static_cast<void(BaseWSPeer<Handler, Impl>::*)(void)>(&BaseWSPeer::close),
```
I suggest a clarifying comment for the need for this case, and a slight simplification in its syntax:
```
            // disambiguate to this close overload
            static_cast<void(BaseWSPeer::*)()>(&BaseWSPeer::close),
```

movitto commented:
@HowardHinnant updated

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

scottschurr commented:
@movitto, congratulations.  It looks like both @intelliot and @HowardHinnant are good with your changes (I'm guessing based on a :+1:).  I think the next step would be to rebase your work on top of the current develop (which is 1.3.0-b4).  That should resolve whatever conflicts there are.  Thanks.

movitto commented:
@scottschurr sure thing will do. That being said, I have a few approved PR's that are outstanding now & I'm a bit concerned about bit rot w/ them, eg the longer these things wait & are in the queue the more likely they'll need to be rebased upon changes pushed to develop (not a big deal for the smaller ones but for #2959 every rebase will be a big undertaking). Perhaps we can establish a merge window in the near future to get these in?

scottschurr commented:
@movitto bit rot on pull requests is a legit concern.  I took a quick glance at your current pull requests:

- https://github.com/ripple/rippled/pull/2959: **Passed**.  I would expect this to be merged to 1.3.0-b4.
- https://github.com/ripple/rippled/pull/2949: **Passed**.  I would expect this to be merged to 1.3.0-b4.
- https://github.com/ripple/rippled/pull/2948: Not yet passed.  Can't be merged with conflicts.
- https://github.com/ripple/rippled/pull/2932: Awaiting review from @nbougalis and @mtrippled.
- https://github.com/ripple/rippled/pull/2931: Blocked indefinitely waiting for IANA.

Your biggest concern for bit rot, https://github.com/ripple/rippled/pull/2959, is very likely to go into 1.3.0-b4.  This pull request, https://github.com/ripple/rippled/pull/2948, looks like it has a good chance of going into 1.3.0-b4 once the conflicts are resolved.  I can't really speak to your remaining non-passed pull requests.

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

movitto commented:
Cool, thx for addressing my concern @scottschurr. Just rebased & pushed.

nbougalis commented:
@manojsdoshi, can you please include this PR in the **rc1** candidate you're putting together? Thanks.

scottschurr commented:
Actually, @manojsdoshi, @seelabs is currently reviewing the asio portion of this pull request and may request some changes.  So please don't include this pull request in rc1.  @seelabs has some obligations today so he's unable to finish the review immediately.  However I expect to see his feedback tomorrow.

@movitto, sorry for the delay.  But I think the more eyes the better.

seelabs commented on src/ripple/rpc/impl/ServerHandlerImp.cpp:
```diff
@@ -388,7 +388,7 @@ ServerHandlerImp::processSession(
     if (is->getConsumer().disconnect())
     {
-        session->close();
+        session->close(1008, "threshold exceeded");
```
Instead of the magic number `1008` here, could we use `boost::beast::websocket::policy_error`?

seelabs commented on src/ripple/server/WSSession.h:
```diff
@@ -140,6 +140,10 @@ struct WSSession
 
+    virtual
+    void
+    close(const std::uint16_t& code, const std::string& reason) = 0;
```
Instead of passing the `code` and `reason`, could be pass `boost::beast::websocket::close_reason` instead?

(also, Nit: We prefer to take integer types like `std::uint16_t` by value instead of `const&`)

seelabs commented on src/ripple/server/impl/BaseWSPeer.h:
```diff
@@ -258,6 +264,33 @@ close()
+            // disambiguate to this close overload (also above)
+            static_cast<void(BaseWSPeer::*)(const std::uint16_t&, const std::string&)>(&BaseWSPeer::close),
+            impl().shared_from_this(),
+            code, reason));
```
Instead of `std::bind` and a `static_cast` to disambiguate the overload, what do you think about using a lambda instead? Something like:
```
    if (!strand_.running_in_this_thread())
        return post(strand_, [self = impl().shared_from_this(), reason] {
            self->close(reason);
        });
```

seelabs commented on src/ripple/server/impl/BaseWSPeer.h:
```diff
@@ -258,6 +264,33 @@ close()
+                  std::placeholders::_1)));
+    }
+}
+
```
This function and the one above are almost identical. The only difference is one passes an empty reason while the other passes the caller given reason. How about having the `close()` method create an empty reason and call the `close(boost::beast::websocket::close_reason const&)` function? (Note: I suggest replacing the `close(std::uint16_t const&, std::string const&)` function with one taking a single `reason` parameter elsewhere in this review).

seelabs commented on src/ripple/server/impl/BaseWSPeer.h:
```diff
@@ -258,6 +264,33 @@ close()
+            code, reason));
+    do_close_ = true;
+    if(wq_.empty()){
+      boost::beast::websocket::close_reason _reason;
```
Nit: we usually don't name variables with a leading underscore.

scottschurr replied:
Additional nit: standard indentation in this code base is four spaces.  The bulk of this pull request adheres to that quite well.  However this particular area, starting what the `if(wq_.empty())` has some indentation issues.

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

movitto commented:
Hey I was able to give @seelabs' followup patch a quick lookover this morning (am travelling & out of town for a few days) and I agree with the proposed changes, the code looks much cleaner. I also gave it a whirl and it looks to cleanly build locally though I didn't get as far as testing/verify the actual functionality (I had previously done this using the example provided in #2883). But outside of that, no objections from me, merge away!

movitto commented:
Speaking of merging is the idea for me to fold @seelabs' patch into mine? I can do this though would be a bit concerned that we'd lose the history of both myself and @seelabs contributing to this one. It's probably NBD for this one, but it's often nice to see all the contributors to patchset in the commit log (incase of questions, issues, etc later on). Just let me know!

And thanks again for the review! (I'll get to the other PR's with outstanding comments early next week)

seelabs commented:
@movitto Great, glad the changes look OK to you. I'd like to get a thumbs up from @miguelportilla before we mark this passed. 

Yes, the idea is for you to merge my changes into yours. Usually the PR owner would cherry pick my commit and add it to this PR (so reviewers can leave comments). When this eventually passes, squash all the commits down to one commit with you as the author.

miguelportilla commented:
@seelabs I've reviewed your patch and left some comments.

With @seelabs patch, this PR looks fine to me. Thank you for the contribution @movitto!

seelabs commented:
@movitto I just pushed another patch to address @miguelportilla's feedback. Can you cherry pick this one as well: https://github.com/seelabs/rippled/commit/2741d6750336ea63719e5e98867423ef3a19c04e (https://github.com/seelabs/rippled/commit/891889c35f49ba937bdcdd10725ab9f8313f8af5 should be cherry picked first).

Of course, this is your PR, so please review the code in the new patch, and if you object to anything let me know. 

@HowardHinnant Can you also review this new patch?

movitto commented:
@seelabs just left some comments on @miguelportilla comments (comment'ception?! :smile: ), just let me know when it's good to squash (back in town, should be fairly quick)

movitto commented:
Ahh looks like our comment collided in mid air! (we didn't even reach cruising altitude!) Will try out the changes and if all works locally squash / repush

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

movitto commented:
@seelabs @miguelportilla @HowardHinnant @scottschurr OK, updated:

- cherry picked @seelabs' follow-on patches and reviewed locally. Everything seems inorder tho TBH I'm not 100% up to speed w/ the semantics of the message queue (wr_) and the relevant state (empty, non-empty, etc). That being said I did try out the example provided in #2883 and re-verified the fix.
- squashed & rebased onto develop
- force pushed

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

ripplelabs-jenkins commented:
**Thank you** for your submission. It will be reviewed soon and submitted for processing in CI.

seelabs commented:
@movitto Thanks!

One quick nit: we prefer not to squash the commits down until the PR is officially passed (it makes it easier to review if the reviews can see in the commits what's new). No big deal that it was squashed now, just wanted to bring it up for next time.

HowardHinnant commented:
Still 👍 from me.

miguelportilla commented:
👍
